File: app/src/main/java/bardurt/com/androidalphaindexer/ModelAdapter.java
package bardurt.com.androidalphaindexer;

import android.app.Activity;
import android.content.Context;
import android.support.annotation.IdRes;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.SectionIndexer;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Set;

/**
 * Created by Bardur on 14/05/2017.
 *
 * Model adapter for Android ListView
 */
public class ModelAdapter extends ArrayAdapter<MyModel> implements SectionIndexer {

    private final HashMap<String, Integer>    mAlphaIndexer;
    private final String[]                    mSections;

    public ModelAdapter(@NonNull Context context, @LayoutRes int resource, @IdRes int textViewResourceId, @NonNull List<MyModel> objects) {
        super(context, resource, textViewResourceId, objects);

        mAlphaIndexer = new HashMap<>();
        int size = objects.size();

        for (int x = 0; x < size; x++) {
            String s = objects.get(x).getName();


            String ch = s.substring(0, 1);

            ch = ch.toUpperCase();

            if (!mAlphaIndexer.containsKey(ch)) {
                mAlphaIndexer.put(ch, x);
            }

        }

        Set<String> sectionLetters = mAlphaIndexer.keySet();

        // create a list from the set to sort
        ArrayList<String> sectionList = new ArrayList<>(
                sectionLetters);

        Collections.sort(sectionList);

        mSections = new String[sectionList.size()];

        sectionList.toArray(mSections);
    }

    @NonNull
    @Override
    public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {

        ModelWrapper wrapper;

        if(convertView==null){
            convertView = ((Activity)getContext()).getLayoutInflater().inflate(R.layout.list_item_model, parent,false);
            wrapper =  new ModelWrapper(convertView);
            convertView.setTag(wrapper);
        }else {

            wrapper = (ModelWrapper) convertView.getTag();
        }

        wrapper.populateFrom(getItem(position));

        return convertView;
    }

    @Override
    public int getViewTypeCount() {
        if (getCount() != 0)
            return getCount();

        return 1;
    }

    @Override
    public int getItemViewType(int position) {

        return position;
    }

    @Override
    public Object[] getSections() {
        return mSections;
    }

    @Override
    public int getPositionForSection(int i) {
        return mAlphaIndexer.get(mSections[i]);
    }

    @Override
    public int getSectionForPosition(int i) {
        return 0;
    }
}

File: app/src/main/java/bardurt/com/androidalphaindexer/ModelAdapter.kt
package bardurt.com.androidalphaindexer

import android.app.Activity
import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.SectionIndexer
import androidx.annotation.IdRes
import androidx.annotation.LayoutRes
import java.util.Locale

class ModelAdapter(
    context: Context,
    @LayoutRes resource: Int,
    @IdRes textViewResourceId: Int,
    objects: List<MyModel>
) : ArrayAdapter<MyModel?>(context, resource, textViewResourceId, objects), SectionIndexer {

    private val mAlphaIndexer = HashMap<String, Int>()
    private val mSections: Array<String>

    init {
        val size = objects.size

        for (x in 0 until size) {
            val s = objects[x].name


            var ch = s.substring(0, 1)

            ch = ch.uppercase(Locale.getDefault())

            if (!mAlphaIndexer.containsKey(ch)) {
                mAlphaIndexer[ch] = x
            }
        }

        val sectionLetters: Set<String> = mAlphaIndexer.keys

        // create a list from the set to sort
        val sectionList = ArrayList(
            sectionLetters
        )

        sectionList.sort()

        mSections = arrayOf()

        sectionList.toArray(mSections)
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var view = convertView
        val wrapper: ModelWrapper

        if (view == null) {
            view = (context as Activity).layoutInflater.inflate(
                R.layout.list_item_model,
                parent,
                false
            )
            wrapper = ModelWrapper(view)
            view.tag = wrapper
        } else {
            wrapper = view.tag as ModelWrapper
        }

        wrapper.populateFrom(getItem(position)!!)

        return view!!
    }

    override fun getViewTypeCount(): Int {
        if (count != 0) return count

        return 1
    }

    override fun getItemViewType(position: Int): Int {
        return position
    }

    override fun getSections(): Array<String> {
        return mSections
    }

    override fun getPositionForSection(i: Int): Int {
        return mAlphaIndexer[mSections[i]]!!
    }

    override fun getSectionForPosition(i: Int): Int {
        return 0
    }
}